### DISCUSSION TITLE
Examples for filtering and pagination?

### DISCUSSION THREAD

**Discussion by frederikhors** [SeaQL/sea-orm#766] [Q&A]
_Congratulations again for the work done so far._

I'm wondering if there is somewhere a more common example of complex pagination and filtering.

Example:

1. "sort on multiple fields",
2. "filter with EQ, NEQ, AND, OR on field"
3. etc.

**Comment by billy1624** [comment]
Hey @frederikhors, thanks for the supports!

For example,

```rs
use sea_orm::{entity::*, query::*, tests_cfg::cake};
let mut cake_pages = cake::Entity::find()
    .filter(cake::Column::Id.gt(10))
    .order_by_asc(cake::Column::Id)
    .paginate(db, 50);
 
while let Some(cakes) = cake_pages.fetch_and_next().await? {
    // Do something on cakes: Vec<cake::Model>
}
```

**Reply by frederikhors** [reply]
Yeah. Thanks, but I know that.

What I can't understand (I'm approaching Rust for the first time) is how to transform a user input (perhaps coming from a graphql query) into code for SeaORM.

So what you wrote generated by a GraphQL query or in any case a simple querystring.

**Reply by billy1624** [reply]
Okay, then you need to parse the user input (GraphQL query) and then apply the filter and sorting on SeaORM select.

Btw... we're developing GraphQL support for SeaORM, stay tuned
- https://github.com/SeaQL/summer-of-code/blob/main/2022/README.md#1-a-graphql-framework-on-top-of-seaorm

**Reply by frederikhors** [reply]
For now I'm using GraphQL in Golang with [gqlgen](https://github.com/99designs/gqlgen) and [Bun](https://github.com/uptrace/bun) as ORM.

I would like to use Rust in the near future and imagine using [Async-Graphql](https://github.com/async-graphql/async-graphql) and [SeaORM](https://github.com/SeaQL/sea-orm).

I'm using [DDD](https://en.wikipedia.org/wiki/Domain-driven_design) and [Clean architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html), meaning that I'm avoiding `one-size-fits-all/lock-in` solutions like Hasura, Postgraphile, DGraph and everything like that; I can (and I eventually will, no hurry) write and organize my code in such a way to achieve their features.

For now I'm wondering about complex queries with filters like: `AND`, `OR`, `NOT` ([better described here](https://github.com/graphile-contrib/postgraphile-plugin-connection-filter/blob/master/docs/examples.md)).

Right now I need to think/write/test code all by myself and this certainly does not speed up production of the software I want to write.

Also, my code will almost certainly be slow, buggy and possibly all wrong as soon as I realize that I will need something extra for this feature.

_After all, this is true for all software and is one of the reasons why open source was born._

After some research I noticed there are two approaches to solving the problem of GraphQL data filtering:

- create all possible combinations for each field of the model ([this is how ent behaves](https://entgo.io/blog/2021/07/01/automatic-graphql-filter-generation) and [other users suggest](https://github.com/99designs/gqlgen/issues/385#issuecomment-432130569))

- create input types that contain the model fields but also logical comparators as a possible value (like [Hasura, Postgraphile and others do](https://hasura.io/docs/latest/graphql/core/databases/postgres/queries/query-filters/))

### QUESTIONS

1. What do you think about? Pros and cons of each choice?

1. If I wanna use `async-graphql` with `sea-orm` is it possible to have an easier way to map our Graphql structs to our domain/repository code without manually write that logic by ourself?

1. This writing of mine summarizes what you are thinking and that you have summarized at the link you sent me, @billy1624?

**Comment by tyt2y3** [comment]
You can look at the current GraphQL example we ship with SeaORM. And yes I think the current state is quite mundane.

As mentioned, we have a Summer of Code project about building GraphQL resolvers with SeaORM, and I think our philosophy is quite similar to ent: strongly typed, schema first, code generation. We'd also appreciate your use case, thoughts and imagination here for what makes a good developer experience.

I am sorry but I don't understand your description about the difference between ent and hasura, but I strongly think that we'd be closer to ent.

Ref: https://github.com/SeaQL/summer-of-code/discussions/12 @karatakis
